### PR TITLE
Git Repository Overhaul

### LINKED ISSUE #154: Make Zarf Deployment Push All Git Repository Branches
Currently, when using Zarf to deploy a git repository only the master branch and tags are pushed to the Gitea mirror. It would be ideal if this mirror also contained the other branches that the original git repository contained.

My approach to this would be as follows:
When provided with a Git repository on package deployment, utilize the git refs in `.git/refs/remotes` to populate the refs in Gitea. This directory, unlike `.git/refs/heads` is updated on `git fetch` rather than a `git checkout`, which Zarf runs after cloning the repository, implying that the git refs are already there and just need to be referenced.

### PR DESCRIPTION
**Pull request by JustinFirsching**
Update the pull function so that tag-provided mirrors do not fetch all
tags, instead later fetching (externally) only the tag they need.

Implement tag fetching function to retrieve _only_ the desired tag when
creating a tag-provided repo mirror.

Checkout the tag into a detatched HEAD state at the end of the create
stage of tag-provided repository mirrors.

Implement ref removal and addition functions used during package
creation and mirror push to ensure that refs that aren't wanted on the
mirror won't be pushed.

Update the refspecs used in the push to push detatched HEAD, branches,
online remote, and tags to the offline mirror. Note that if we later
checkout a branch from the remote and do not clean up the remote ref it
will lead to a duplicate ref name and the push will fail on one of the
refs (likely the online one since it is later in the refspec slice).

Fixes #154 

feat: Allow for repos to be provided without a tag to mirror all branches/tags
feat: Make tag-provided repository mirrors use the tag as master
fix: Prevent tag-provided repo mirrors from storing extra refs

### CODE REVIEW

**Comment by RothAndrew** [comment]
I'll let @jeff-mccoy be the final say here but I don't think this is something we want to do. We want the declarative state to be a specific git reference. Its an outstanding bug that `zarf package create` actually clones more than is intended.

**Comment by JustinFirsching** [comment]
Responded in the issue as well, but to make sure it is seen here too:

Would it be fair to say that if there is no tag provided everything is brought over, but if a tag is provided only that tag is brought over? If so, I could likely get the rest of this implemented this week.

**Comment by jeff-mccoy** [comment]
I commented on the issue, open to discussing this as an opt-in, but the reason for tag-based tracking is we will be using that to do diffs in the future without having to fetch each remote asset and see if it changed.  In git tags can’t be moved, they can only be removed/ added (ie forced) but that’s not the standard convention so it’s a good way to track changes quickly.  Long term idea would be a new command to compare a current zarf.yaml to an older one to see what resources have changed and only include those.

**Comment by jeff-mccoy** [comment]
@YrrepNoj please review for code flow.  @RothAndrew please review for overall user flow and how this impacts current examples such as the Big Bang example.

**Comment by jeff-mccoy** [comment]
I’ll do a final review after both of you.

**Comment by RothAndrew** [comment]
@JustinFirsching pls pull in latest master changes

**Comment by JustinFirsching** [comment]
@RothAndrew Sorry about that. Rebased and pushed.

**Comment by RothAndrew** [comment]
No worries, just wanted to make sure the stuff added in 0.13 got included, namely the change to the private registry to require auth.

File: cli/internal/packager/create.go
Comment (YrrepNoj) [inline cli/internal/packager/create.go:125]: I'm fine with it either way but I think this might make a lot of sense to put at the end of the `git.Pull()` call.

**Review by YrrepNoj — APPROVED** [review]

**Review by RothAndrew — CHANGES_REQUESTED** [review]
- When deploying the `gitops-service-data` example based on the zarf.yaml I expect to see 3 repos, but I only see 1.
image

- Recommend having the `gitops-service-data` example pull in a helm chart that gets deployed. I see the Twistlock chart getting pulled in in the zarf.yaml but I don't really have any good way to verify that it is there
- examples/gitops-data/README.md should be updated to talk about the different behaviors being demoed (single tag & full repo)
- IMO we should add an E2E test for the `gitops-service-data` example but I won't hold up the PR for it if everyone else just wants to keep pushing forward

File: cli/internal/packager/create.go
Comment (YrrepNoj) [inline cli/internal/packager/create.go:125]: I'm with with it either way but I think this logic would make sense to put at the end of the `git.Pull( ... )` function.

Comment (JustinFirsching) [reply cli/internal/packager/create.go:125]: Actually, the first implementation I had for this did look like that haha. I ended up switching to this because I wasn't sure of the scope of git pull usage for Zarf. I figured that it would be easier to implement it with no limitations (such as deleting the refs) so that if a pull were ever needed for a different reason, that didn't fetch all tags, a refactor wasn't needed. Hopefully that at least clears up my thought process for not including it directly in the pull. If `git.Pull` will only ever be needed for charts pulls and repo mirrors, I see no reason not include the tag checkout/reference deletion logic at the end, and if you'd like, I can make that happen while handling the README update.

**Comment by JustinFirsching** [comment]
@RothAndrew 

- Is it possible you are using an old copy of the Zarf package? I am getting all three repositories as shown in the screenshot attached. Running `tar -I unzstd -tf zarf-package-gitops-service-data.tar.zst | grep 'mirror__[^/]*/.git/$'` should list all of the repos in the package, and I expect the tar you have will only have the one repository.
image
- Would I just need to deploy the chart using a Kubernetes manifest? If so, I think that would look like joining the current (master) version of `examples/gitops-data` with `examples/single-big-bang-package`, which I can do, but want to verify before creating a duplicate code chunk.
- As for the README.md, I won't have a chance to get around to that tonight, but can definitely get that updated to explain the new features. 
- Assuming all that needs to be done for the E2E test is updating the `e2e.sh` script to include a few calls to the Gitea API confirming that a gitops-data deployment pushed the repositories as expected, I can toss that in with the README.md updates.

**Comment by RothAndrew** [comment]
Its possible. Ill check again im the morning

**Comment by RothAndrew** [comment]
e2e.sh is the old way that we are moving away from. The new way is golang tests with the Terratest library, in the /test directory. If you feel comfortable taking a stab at it thats great. If not that's okay too, we can make an issue to follow up with a new test

**Comment by RothAndrew** [comment]
> Is it possible you are using an old copy of the Zarf package? I am getting all three repositories as shown in the screenshot attached. Running tar -I unzstd -tf zarf-package-gitops-service-data.tar.zst | grep 'mirror__[^/]*/.git/$' should list all of the repos in the package, and I expect the tar you have will only have the one repository.

Yep, I had `master` checked out rather than the branch you're working in. In your branch I see all 3 repos, and everything looks good inside them. The Zarf repo only has master (at the point in time when 0.12.0 was cut) and the 0.12.0 tag, and podinfo and twistlock have everything 👍 

> Would I just need to deploy the chart using a Kubernetes manifest? If so, I think that would look like joining the current (master) version of examples/gitops-data with examples/single-big-bang-package, which I can do, but want to verify before creating a duplicate code chunk.

IMO I'd actually just take out the chart from the `gitops-data` example. That functionality is already demoed in a couple of other places, including the `single-big-bang-package` example.

> As for the README.md, I won't have a chance to get around to that tonight, but can definitely get that updated to explain the new features.

👍 

> Assuming all that needs to be done for the E2E test is updating the e2e.sh script to include a few calls to the Gitea API confirming that a gitops-data deployment pushed the repositories as expected, I can toss that in with the README.md updates.

See above comment from last night

**Comment by JustinFirsching** [comment]
Updated the README as a new commit to make it clear that nothing else was changed. Let me know what still needs to be changed and I can rebase from there. 

I haven't looked into the E2E tests much, but I don't have an AWS account, so if I understand correctly that may rule me out entirely.

**Comment by RothAndrew** [comment]
> Updated the README as a new commit to make it clear that nothing else was changed. Let me know what still needs to be changed and I can rebase from there.
> 
> I haven't looked into the E2E tests much, but I don't have an AWS account, so if I understand correctly that may rule me out entirely.

The tests would run in our account, they'd just need to be written and wired up, but I don't necessarily expect everyone to do that, at least until we write a guide on how to do it.

This can be merged without changes to the E2E test suite with an issue to have one of the core devs to do it.

File: examples/gitops-data/README.md
Comment (RothAndrew) [inline examples/gitops-data/README.md:25]: If the trunk branch of my repo is named `main` would it still show up as `master` in Gitea?

Comment (JustinFirsching) [reply examples/gitops-data/README.md]: It would have, recent changes correct this behavior.

File: cli/internal/git/pull.go
Comment (jeff-mccoy) [inline cli/internal/git/pull.go:39]: Can you clarify this comment?  I think I understand what you are saying in that we now support with and with the tag, but to be clear the default and normal behavior we will document/encourage will use the tag as that will be very important for upcoming diff engine work

File: cli/internal/git/pull.go
Comment (jeff-mccoy) [inline cli/internal/git/pull.go:71]: I redid this a little, I've had issue in golang before doing `err == nil` and try to always use `err != nil` because of past issues, https://yourbasic.org/golang/gotcha-why-nil-error-not-equal-nil/.  The nested conditionals were a little hard to follow.  Most of this is subjective, but I always strive to reduce cognitive load when possible.  Feel free to change some more if I messed this up.

File: cli/internal/git/pull.go
Comment (JustinFirsching) [reply cli/internal/git/pull.go:39]: Good catch! That comment was placed when I did the first commit to switch to full mirrors and although I have changed the variable below it since, I missed deleting the comment. This comment needs to be deleted. I can push a deletion commit now, or include one when I force push a squash. I've just been keeping everything as separate commits lately so reviewing changes such as tag branch name targets could be done in steps, which I assume is how it has been missed until now.

File: cli/internal/git/pull.go
Comment (JustinFirsching) [reply cli/internal/git/pull.go:71]: That looks good to me. I wasn't a huge fan of the nested conditional either, was just trying to avoid this:
```go
if err != nil {
    logContext.Debug(err)
    logContext.Warn(message)
} else if !head.Name().IsBranch() {
    logContext.Warn(message)
} else {
    trunkBranchName = head.Name()
}
```

since the `if` and `else if` warning log are the same message. With how you have the messages setup I like this much better.

Comment (JustinFirsching) [reply cli/internal/git/pull.go:71]: That's very interesting with the whole `nil != nil` thing. Thanks for the share!

File: cli/internal/git/utils.go
Comment (jeff-mccoy) [inline cli/internal/git/utils.go:245]: Not super clear on this, the negative logic is a little awkward, can we do `ref.Name() != plumbing.HEAD && ref.Name() != head.Name()` instead?  The name passHead isn't very clear either.  Maybe a little more comments on what is happening here?

**Review by jeff-mccoy — COMMENTED** [review]
Overall this looks good, some minor comments on some conditionals that are a little unclear (I patched one of them already) and look to @RothAndrew's final review as well.  Think we're very close.  Thanks for the work on this, it will make git operations much more robust.

File: cli/internal/git/pull.go
Comment (jeff-mccoy) [reply cli/internal/git/pull.go:39]: either way we'll just need all the commits in before final test/review, up to you, thanks.

File: cli/internal/git/pull.go
Comment (jeff-mccoy) [reply cli/internal/git/pull.go:71]: Yeah I had to play with the logic a little to get it how I thought it might read better.  I also did different warns since technically they are different conditions occurring between the err and not being a branch.

File: cli/internal/git/utils.go
Comment (JustinFirsching) [reply cli/internal/git/utils.go:245]: Sure! 

To explain, I called it `passHead` because it determines if we pass the HEAD check **for deletion**, which is `and`ed with if we pass the user check for deletion. I can find a better name for it though since `passHead` certainly isn't very descriptive on its own. As for the awkward logic, I wrote it that way to reflect "what would a HEAD check look like", and then "now invert that since we want to pass the HEAD check when the ref is _not_ HEAD or what HEAD points to". Refactoring to "what implies that the ref isn't HEAD or what HEAD points to" makes sense to me though. I'll add comments when I do so as well.

**Comment by JustinFirsching** [comment]
Thanks! I will get the requested changes up as soon as possible.

Looking forward to the final review. I assume it would be easiest to review as separate commits since @RothAndrew has been tracking along the way, but if that is not the case, let me know and I can force push with the commits squashed.

This was a much more interesting and complex issue than I expected it to be when I first filed the issue and I'm glad to have been able to contribute so much of the solution.

**Comment by RothAndrew** [comment]
You can go ahead and squash if you want. Im off tomorrow but I'll take a look if I have some down time. I just had a couple of nitpicks on the documentation that may have been resolved now. @jeff-mccoy if you feel good about it thats good enough for me

**Comment by jeff-mccoy** [comment]
I’d say press with the squash and run a “/test all” @JustinFirsching

**Comment by jeff-mccoy** [comment]
@JustinFirsching once the review comments I made a couple days ago are addressed I think we can proceed to squash/merge this PR.

**Comment by JustinFirsching** [comment]
Sorry for the delay on this! Made the changes and got sidetracked before I had a chance to rebase and push. I don't think I have permissions to run tests, so someone else may need to kick those off for me, but I think this is what we are looking for. Happy to resolve any remaining issues with the changes and/or documentation associated with it!

File: examples/gitops-data/README.md
Comment (jeff-mccoy) [inline examples/gitops-data/README.md:74]: good news with the native apply changes this wont be an issue anymore

**Review by jeff-mccoy — APPROVED** [review]
everything looks good, thanks for all the work on this!

**Comment by JustinFirsching** [comment]
Glad to contribute! Thank you all for the feedback and being so responsive on the PR and issue thread!

**Pull request merged**
